src/eos-updater-apply.c

```diff
@@ -73,6 +74,431 @@ get_test_osname (void)
+}
+
+static gboolean
+pull_flatpaks (GPtrArray    *flatpaks_to_deploy,
+               GCancellable *cancellable,
+               GError       **error)
+{
+  g_autoptr(FlatpakInstallation) installation = get_flatpak_installation (cancellable,
+                                                                          error);
+  gsize i = 0;
+
+  if (!installation)
+    return FALSE;
+
+  for (; i < flatpaks_to_deploy->len; ++i)
+    {
+      FlatpakToInstall *to_install = g_ptr_array_index (flatpaks_to_deploy, i);
+      g_autoptr(GError) local_error = NULL;
+
+      /* We have to pass in a local_error instance here and check to see
+       * if it was FLATPAK_ERROR_NOT_INSTALLED - this is what will be
+       * thrown if we succeeded at pulling the flatpak into the local
+       * repository but did not deploy it (since no FlatpakInstalledRef
+       * will be returned). */
+      flatpak_installation_install_full (installation,
+                                         FLATPAK_INSTALL_FLAGS_NO_DEPLOY,
+                                         to_install->repo,
+                                         to_install->ref_kind,
+                                         to_install->app_identifier,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         cancellable,
+                                         &local_error);
+
+      /* This only happens if the wrong version of flatpak is deployed
+       * with eos-updater */
+      if (!local_error)
+        {
+          g_critical ("Flatpak installation should not have succeeded!");
+          g_assert (FALSE);
+        }
+
+      /* Something unexpected failed, return early now.
+       *
+       * We are not able to meaningfully clean up here - the refs will remain
+       * in the flatpak ostree repo for the next time we want to install them
+       * but there's nothing in the public API that will allow us to get rid of
+       * them. */
+      if (!g_error_matches (local_error, FLATPAK_ERROR, FLATPAK_ERROR_NOT_INSTALLED))
+        return FALSE;
+    }
+
+  return TRUE;
+}
+
+static const gchar *
+get_pending_flatpak_deployments_state_path (void)
+{
+  return eos_updater_get_envvar_or ("EOS_UPDATER_TEST_UPDATER_FLATPAK_UPGRADE_STATE_DIR",
+                                    LOCALSTATEDIR "/lib/eos-updater");
+}
+
+static gboolean
+write_flatpaks_to_be_deployed (GPtrArray    *flatpaks_to_deploy,
+                               GCancellable *cancellable,
+                               GError       **error)
+{
+  const gchar *pending_flatpak_deployments_state_path = get_pending_flatpak_deployments_state_path ();
+  const gchar *pending_flatpak_deployments_path = g_build_filename (pending_flatpak_deployments_state_path,
```

Comment:
Should this path be made specific to the just-written deployment’s checksum? I’d like to avoid the situation where someone upgrades their system, reboots, but reboots back into the old deployment (which could happen in future if the new deployment fails to boot [perhaps due to kernel problems, for example]) and yet ends up deploying flatpaks which are not intended for that deployment.

Reply:
That's a good point. I did consider making it deployment-checksum-specific though I guess a side effect of that is that you're going to leave a lot of litter around `/var` (I'd imagine we'd have something like `/var/lib/eos-updater/to-export/csum`)

Reply:
The litter would occur in the "failed to boot deployment" case. Though I guess we could treat those `to-export` directories as "one-time" to-export commands for each deployment.

Reply:
When cleaning up the `to-export` directory, you could delete any entries where there is no longer an OSTree deployment with a matching checksum. That basically ties in the cleanup here to OSTree’s deployment cleanup mechanism.

Reply:
That could also work